### ISSUE TITLE
Assert.DoesNotThrow() stopped working as it was previously

### ISSUE THREAD

**Issue by carbon88** [nunit/nunit#3315]
Some specs of environment first.
Windows 10 1809 build 17763.557
VS 2017 v15.9.12
.Net Framework Version: 4.7.03190
Test project is targeted for .Net Core 2.0.3
NUnit package version: 3.12.0 (latest for now)
NUnit3TestAdapter package version 3.12.0

Let's say we have an extension method like:

```
public static void Run<T>(this IEnumerable<T> sequence)     {
     foreach (var item in sequence);
}
```

and I used to use it like
  `Assert.DoesNotThrow(someEnumerable.Run)`

because IDE (or resharper, whatever)  suggests me use it instead of 
  `Assert.DoesNotThrow(()=>someEnumerable.Run())`

And everything was fine but then I had to update because another bug was fixed it started failing my tests with the exception
"System.ArgumentException : The actual value must be a parameterless delegate but was TestDelegate.
Parameter name: parameterName"

What is that? Do you even test the changes you make, guys?

Here is the test:

```
[Test]
public void NunitDoesNotThrowTest()
{
  TestDelegate fullform = () => Enumerable.Repeat(1, 10).Run(); // it is your TestDelegate. working
  TestDelegate shortForm = Enumerable.Repeat(1, 10).Run; // this is to. should be working because IT IS your TestDelegate (acording to VS) you want me to pass into DoesNotThrow() method!
  Assert.DoesNotThrow(() => Assert.DoesNotThrow(fullform));
  Assert.DoesNotThrow(() => Assert.DoesNotThrow(shortForm));
}
```

**Comment by mikkelbu** [comment]
I've looked a bit into this and the problem occurs when we have a combination of method groups and extension methods, so that is probably why nobody has hit this before. E.g. the following works

```csharp
[Test]
public static void NunitDoesNotThrowTest3()
{
    TestDelegate nonExtensionMethod = Run;
    Assert.DoesNotThrow(nonExtensionMethod);
}

internal static void Run()
{ }
```

I'm unsure whether the right approach will be to special case `TestDelegate` in `RecordException` or whether we can detect that we are calling an extension method. I'm leaning towards the first solution.

**Comment by carbon88** [comment]
Well, it is not against TestDelegate signature, right? no incoming parameters no returning type just an operation. it could have closure variables access but in full form it works fine. Maybe it is because of closure? I didn't go deeper with it.

Maybe if you really need full form you can just transform `Enumerable.Repeat(1, 10).Run;` into `() => Enumerable.Repeat(1, 10).Run();` internally so people continue working how they are used to and  don't have to rewrite something what was fine in hundreds of tests.

**Comment by jnm2** [comment]
Looks like a clear bug, easy to fix. This one was on me. Until a minute ago, I had no idea that C# permitted that syntax with an extension method.

> What is that? Do you even test the changes you make, guys?

What, testing? That doesn't sound like us... 😉

**Comment by jnm2** [comment]
@carbon88 It was what I thought (https://github.com/nunit/nunit/pull/3317). Sorry you ran into this bug!

**Comment by k15tfu** [comment]
@jnm2 Hi! Any chance to get it fixed in 3.12.1? Not sure when 3.13 will be released.

**Comment by rprouse** [comment]
3.13 will be released soon. We don't do point releases for features, only for serious bugs or regressions. If you want to use this feature now, you can always use our continuous builds.
